Mathlib/Combinatorics/CombinatorialMap.lean

```diff
@@ -0,0 +1,187 @@
+permutations.
+-/
+def Hom (M : CombinatorialMap D) (M' : CombinatorialMap D') (f : D → D') : Prop :=
+  f ∘ M.σ = M'.σ ∘ f ∧
+  f ∘ M.α = M'.α ∘ f ∧
+  f ∘ M.φ = M'.φ ∘ f
+
+private lemma hom_inv_is_hom_aux {p₁ : Equiv.Perm D} {p₂ : Equiv.Perm D'} {f : Equiv D D'}
+    (h : f ∘ p₁ = p₂ ∘ f) : f.symm ∘ p₂ = p₁ ∘ f.symm :=
+  calc f.symm ∘ p₂ = f.symm ∘ p₂ ∘ id := rfl
+    _ = f.symm ∘ p₂ ∘ (f ∘ f.symm) := by rw [← Equiv.self_comp_symm]
+    _ = f.symm ∘ (p₂ ∘ f) ∘ f.symm := rfl
+    _ = f.symm ∘ (f ∘ p₁) ∘ f.symm := by rw [← h]
+    _ = (f.symm ∘ f) ∘ p₁ ∘ f.symm := rfl
+    _ = id ∘ p₁ ∘ f.symm := by rw [Equiv.symm_comp_self]
+
+/-- The inverse of a homomorphism between `CombinatorialMap`s is itself a homomorphism. -/
+lemma hom_inv_is_hom (M : CombinatorialMap D) (M' : CombinatorialMap D') (f : Equiv D D')
+    (h : Hom M M' f) : Hom M' M f.symm :=
+  ⟨hom_inv_is_hom_aux h.1, hom_inv_is_hom_aux h.2.1, hom_inv_is_hom_aux h.2.2⟩
+
+/-- The opposite of a `CombinatorialMap` which reverses the identities of the darts on each edge. -/
+def opp (M : CombinatorialMap D) : CombinatorialMap D where
+  σ := M.α * M.σ * M.α
+  α := M.α
+  φ := M.α * M.φ * M.α
+  involutive := M.involutive
+  fixedPoints_isEmpty := M.fixedPoints_isEmpty
+  composition := by
+    ext x
+    simp only [Equiv.Perm.coe_mul, Function.comp_apply, Equiv.Perm.coe_one, id_eq]
+    rw [M.involutive, M.involutive, composition'_apply]
+
+lemma opp_equiv (M : CombinatorialMap D) : ∃ (f : Equiv D D), Hom M M.opp f :=
+  ⟨M.α, ⟨by rw [← (M.α ∘ M.σ).comp_id, ← Function.RightInverse.id M.involutive]; rfl, rfl,
+  by rw [← (M.α ∘ M.φ).comp_id, ← Function.RightInverse.id M.involutive]; rfl⟩⟩
+
+/-- The dual of a `CombinatorialMap` swaps the roles of vertices and faces -/
+def dual (M : CombinatorialMap D) : CombinatorialMap D where
+  σ := M.φ⁻¹
+  α := M.α⁻¹
+  φ := M.σ⁻¹
+  composition := by
+    ext x
+    apply_fun M.α * M.φ * M.σ
+    simp [composition'_apply]
+  involutive := by
+    have : M.α⁻¹ = M.α := M.involutive.symm_eq_self_of_involutive
+    simp [this, M.involutive]
+  fixedPoints_isEmpty := by
+    have : M.α⁻¹ = M.α := M.involutive.symm_eq_self_of_involutive
+    simp [this, M.fixedPoints_isEmpty]
+
+/-- The double dual of a `CombinatorialMap` is equal to the original map. -/
+lemma double_dual_eq (M : CombinatorialMap D) : M.dual.dual = M := rfl
+
+/-- The vertices of a `CombinatorialMap`. -/
+@[reducible] def vertices (M : CombinatorialMap D) :=
+  Quotient (Equiv.Perm.SameCycle.setoid M.σ)
+
+/-- The edges of a `CombinatorialMap`. -/
+@[reducible] def edges (M : CombinatorialMap D) :=
+  Quotient (Equiv.Perm.SameCycle.setoid M.α)
+
+/-- The faces of a `CombinatorialMap`. -/
+@[reducible] def faces (M : CombinatorialMap D) :=
+  Quotient (Equiv.Perm.SameCycle.setoid M.φ)
+
+/-- The vertex corresponding to a dart `d`. -/
+@[reducible] def dartVertex (M : CombinatorialMap D) (d : D) : M.vertices :=
```

Comment:
I think it makes more sense to call this `Vertex.mk`. Similarly for the edges and faces.